src/main/scala/com/databricks/spark/avro/avroRelation2/AvroRelation2.scala

```diff
@@ -0,0 +1,216 @@
+   *
+   * @since 1.4.0
+   */
+  override def dataSchema: StructType = maybeDataSchema match {
+    case Some(structType) => structType
+    case None => SchemaConverters.toSqlType(avroSchema()).dataType.asInstanceOf[StructType]
+  }
+
+
+  /**
+   * Prepares a write job and returns an [[OutputWriterFactory]].  Client side job preparation can
+   * be put here.  For example, user defined output committer can be configured here
+   * by setting the output committer class in the conf of spark.sql.sources.outputCommitterClass.
+   *
+   * Note that the only side effect expected here is mutating `job` via its setters.  Especially,
+   * Spark SQL caches [[BaseRelation]] instances for performance, mutating relation internal states
+   * may cause unexpected behaviors.
+   *
+   * @since 1.4.0
+   */
+  override def prepareJobForWrite(job: Job): OutputWriterFactory = {
+    val build = SchemaBuilder.record(recordName).namespace(recordNamespace)
+    val outputAvroSchema = SchemaConverters.convertStructToAvro(dataSchema, build, recordNamespace)
+    AvroJob.setOutputKeySchema(job, outputAvroSchema)
+    val AVRO_COMPRESSION_CODEC = "spark.sql.avro.compression.codec"
+    val AVRO_DEFLATE_LEVEL = "spark.sql.avro.deflate.level"
+    val COMPRESS_KEY = "mapred.output.compress"
+
+    sqlContext.getConf(AVRO_COMPRESSION_CODEC, "snappy") match {
+      case "uncompressed" =>
+        logInfo("writing Avro out uncompressed")
+        job.getConfiguration.setBoolean(COMPRESS_KEY, false)
+      case "snappy" =>
+        logInfo("using snappy for Avro output")
+        job.getConfiguration.setBoolean(COMPRESS_KEY, true)
+        job.getConfiguration.set(AvroJob.CONF_OUTPUT_CODEC, DataFileConstants.SNAPPY_CODEC)
+      case "deflate" =>
+        val deflateLevel = sqlContext.getConf(
+          AVRO_DEFLATE_LEVEL, Deflater.DEFAULT_COMPRESSION.toString).toInt
+        logInfo(s"using deflate: $deflateLevel for Avro output")
+        job.getConfiguration.setBoolean(COMPRESS_KEY, true)
+        job.getConfiguration.set(AvroJob.CONF_OUTPUT_CODEC, DataFileConstants.DEFLATE_CODEC)
+        job.getConfiguration.setInt(AvroOutputFormat.DEFLATE_LEVEL_KEY, deflateLevel)
+      case unknown: String => sys.error(s"Unknown output compression: $unknown")
+    }
+    new AvroOutputWriterFactory(dataSchema, recordName, recordNamespace)
+  }
+
+
+  /**
+   * This filters out unneeded columns before converting into the internal row representation.
+   * The first record is used to get the sub-schema that contains only the requested fields,
+   * this is then used to generate the field converters and the rows that only
+   * contain `requiredColumns`
+   */
+  override def buildScan(requiredColumns: Array[String], inputs: Array[FileStatus]): RDD[Row] = {
+    if (inputs.isEmpty) {
+      sqlContext.sparkContext.emptyRDD[Row]
+    } else {
+      inputs.map(path =>
+        sqlContext.sparkContext.hadoopFile(
```

Comment:
This is going to result in broadcasting the hadoop conf for each partition.  If you construct the HadoopRDD by hand you could broadcast the configuration once instead.  This would work well a single partition with a lot of files, but not for many partitions.  We'll have to open up some other interfaces in spark to make the latter case more efficient.